// gzipper/src/compressions/Compression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Compression = /** @class */ (function () {
    /**
     * Creates an instance of Compression.
     */
    function Compression(options, logger) {
        this.compressionOptions = {};
        this.options = options;
        this.logger = logger;
        this.selectCompression();
    }
    /**
     * Returns human-readable compression options info.
     */
    Compression.prototype.readableOptions = function (keyWrapper) {
        if (keyWrapper === void 0) { keyWrapper = function (key) { return key; }; }
        var options = '';
        for (var _i = 0, _a = Object.entries(this.compressionOptions); _i < _a.length; _i++) {
            var _b = _a[_i], key = _b[0], value = _b[1];
            options += keyWrapper(key) + ": " + value + ", ";
        }
        return this.compressionName + " -> " + options.slice(0, -2);
    };
    /**
     * Build compression options object [compressionOptions].
     */
    Compression.prototype.selectCompression = function () {
        var options = {};
        if (this.options.level !== undefined) {
            options.level = this.options.level;
        }
        if (this.options.memoryLevel !== undefined) {
            options.memLevel = this.options.memoryLevel;
        }
        if (this.options.strategy !== undefined) {
            options.strategy = this.options.strategy;
        }
        this.compressionOptions = options;
    };
    return Compression;
}());
exports.Compression = Compression;
